packages/transformers/js/src/JSTransformer.js

```diff
@@ -124,76 +129,82 @@ export default new Transformer({
           pkg.browser.fs === false;
 
         if (!ignore) {
-          traverse(ast.program, fsVisitor, null, {asset, logger});
+          traverse(ast.program, fsVisitor, null, {asset, logger, ast});
         }
       }
 
       // Insert node globals
-      if (GLOBAL_RE.test(code)) {
+      if (code == null || GLOBAL_RE.test(code)) {
         asset.meta.globals = new Map();
         walkAncestor(ast.program, insertGlobals, asset);
       }
     }
 
     if (asset.env.scopeHoist) {
-      hoist(asset);
+      hoist(asset, ast);
     } else if (asset.meta.isES6Module) {
       // Convert ES6 modules to CommonJS
-      let res = await babelCore.transformFromAstAsync(ast.program, code, {
-        code: false,
-        ast: true,
-        filename: asset.filePath,
-        babelrc: false,
-        configFile: false,
-        plugins: [require('@babel/plugin-transform-modules-commonjs')],
-      });
-
-      ast.program = res.ast;
-      ast.isDirty = true;
-    }
-
-    return [asset];
-  },
-
-  async generate({asset, options}) {
-    let code = await asset.getCode();
-    let res = {
-      code,
-    };
-
-    let ast = asset.ast;
-    if (ast && ast.isDirty !== false) {
-      let sourceFileName: string = relativeUrl(
-        options.projectRoot,
-        asset.filePath,
-      );
-
-      let generated = generate(
+      let res = await babelCore.transformFromAstAsync(
         ast.program,
+        code ?? undefined,
         {
-          sourceMaps: options.sourceMaps,
-          sourceFileName: sourceFileName,
+          code: false,
+          ast: true,
+          filename: asset.filePath,
+          babelrc: false,
+          configFile: false,
+          plugins: [require('@babel/plugin-transform-modules-commonjs')],
         },
-        code,
       );
 
-      res.code = generated.code;
-      // $FlowFixMe...
-      res.map = new SourceMap(generated.rawMappings, {
-        [sourceFileName]: null,
+      asset.setAST({
+        type: 'babel',
+        version: '7.0.0',
+        program: res.ast,
       });
     }
 
-    if (asset.meta.globals && asset.meta.globals.size > 0) {
-      res.code =
-        Array.from(asset.meta.globals.values())
-          .map(g => (g ? g.code : ''))
-          .join('\n') +
-        '\n' +
-        res.code;
-    }
-    delete asset.meta.globals;
+    return [asset];
+  },
 
+  generate({asset, ast, options}) {
+    let sourceFileName: string = relativeUrl(
+      options.projectRoot,
+      asset.filePath,
+    );
+
+    let generated = generate(
+      ast.program,
+      {
+        sourceMaps: options.sourceMaps,
+        sourceFileName: sourceFileName,
+      },
+      '',
+    );
+
+    let res = {
+      code: generated.code,
+      map:
+        generated.rawMappings != null
+          ? new SourceMap(generated.rawMappings, {
+              [sourceFileName]: null,
+            })
+          : null,
+    };
+
+    res.code = generateGlobals(asset) + res.code;
```

Comment:
So the globals are only inserted when code is generated? I guess this should rather be done at the end of transform?

Reply:
I think this may be to avoid dirtying the AST to save on performance if we're not scope hoisting — we can just reuse the existing code and map. @devongovett?

Reply:
Hmm, does this currently mean that if you have globals but the AST is never dirtied, they are never inserted?

Reply:
Good point. Maybe we should attach the globals as AST nodes during transformation if we're scope hoisting (although I'd prefer not to rely on that), and otherwise continue to add them here in `generate`?

As I understand, we do this to avoid dirtying the AST to prevent generation. The simplest solution would just to prepend the globals during transform as @mischnic suggests. I don't expect most modules to require node globals, so maybe this is worth the perf cost. What do you think @devongovett?

Reply:
Inserting the globals as a string here also probably offsets the sourcemap.